Add HTML meta tag "description" with QGIS version

pathmapper opened the pull request:
ref https://lists.osgeo.org/pipermail/qgis-developer/2023-March/065647.html

from https://github.com/qgis/QGIS-Documentation/issues/6449#issuecomment-759365309

> The missing version in Google result is problematic imho. Is there something we can/need configure on our side?

There's a good chance that Google/other search engines would use an HTML meta tag `description` for the snippet when displaying search results:
https://developers.google.com/search/docs/appearance/snippet#meta-descriptions

---

**This PR**
![image](https://user-images.githubusercontent.com/20856381/226925819-6cab357d-567f-4281-bfcc-9f2b454fa634.png)

---

Here's an example which demonstrates this: https://www.google.com/search?q=geoportal+niederrhein

![image](https://user-images.githubusercontent.com/20856381/226919346-04e6ff60-af8d-4d01-8e66-30e82b452349.png)

Source of https://geoportal-niederrhein.de/
![image](https://user-images.githubusercontent.com/20856381/226919798-79b3c0bf-a158-419e-8f46-fb2944485de3.png)

---

I'm not familiar with the release workflow of the docs, but looking at the code I guess the specific version will be used here when building the docs for deployment:
https://github.com/qgis/QGIS-Documentation/blob/f8b046aefea977e7e7d5386e26aaaa0a2b5b18c5/conf.py#L49

Would be great if someone could confirm :smile: 

---

We could potentially add additional HTML meta tags there, e.g. to exclude outdated docs from search results.

Here are some options:
- https://developers.google.com/search/docs/crawling-indexing/special-tags
- https://developers.google.com/search/docs/crawling-indexing/robots-meta-tag#directives

/cc @DelazJ @rduivenvoorde

Gustry commented on conf.py:
```diff
@@ -148,7 +148,9 @@
-    'isTesting': True
+    'isTesting': True,
+    # HTML meta tags for SEO https://developers.google.com/search/docs/crawling-indexing/special-tags
+    'metatags': '<meta name="description" content="Documentation for QGIS {}" />'.format(version)
```
```suggestion
    'metatags': f'<meta name="description" content="Documentation for QGIS {version}" />',
```

Gustry replied:
I think we can switch safely to some fstring, more readable IMHO ;-)

pathmapper replied:
Thanks Gustry, sure, there are some more in https://github.com/qgis/QGIS-Documentation/blob/master/conf.py :-)

Gustry replied:
I know there are some more, it's just to migrate step by step ;-)
I was just adding a comma for better git diff/blame, when someone add a line, but no worries.

nicogodet replied:
@pathmapper Trailing comma is recommended by PEP 8 (see https://peps.python.org/pep-0008/#when-to-use-trailing-commas)

pathmapper replied:
OK, convinced :-) added it back in https://github.com/qgis/QGIS-Documentation/pull/8121/commits/361e40469761a1585d08f55d7b776e5d4417fb09
Thanks both.

DelazJ commented:
@pathmapper Thanks for addressing this issue
I downloaded the artifact from github action and couldn't find the meta name info as shown in your screenshot above. Still get "docutils...". Any idea?
Anyway, TBH I'm a bit far from this topic and can't find time for the literature you point to (sorry!). But I'm happy to merge and see, if some more versed watcher gives a "GO!"

pathmapper commented:
> I downloaded the artifact from github action and couldn't find the meta name info as shown in your screenshot above. Still get "docutils...". Any idea?

Good find @DelazJ ! Looks like we were running into https://github.com/readthedocs/sphinx_rtd_theme/issues/469.

I've implemented it another way in https://github.com/qgis/QGIS-Documentation/pull/8121/commits/d7a8e7fd5abbea7c5c5a763378b9ebd33446442d and added also the page title to the content to have unique descriptions ([which is recommended by Google](https://developers.google.com/search/docs/appearance/snippet#create-unique-descriptions-for-each-page-on-your-site)).

![image](https://user-images.githubusercontent.com/20856381/227575084-23fb78f0-b851-4876-b1b8-0ebe3c90ba20.png)

> But I'm happy to merge and see, if some more versed watcher gives a "GO!"

I think we have nothing to loose here - well, only the currently automatically generated snippets :-)
But if we loose them, we are going to have a snippet including the version number. :crossed_fingers:

pathmapper commented on themes/rtd_qgis/layout.html:
```diff
@@ -1,5 +1,9 @@
 
+ {%- block extrahead %}
+  <meta name="description" content="Documentation for QGIS {{version}} - {{ title|striptags|e }}">
+ {%- endblock -%}
```
```suggestion
  {%- if outdated %}
  <meta name="robots" content="noindex">
  {%- endif %}
 {%- endblock -%}
```

DelazJ replied:
@pathmapper Committing this suggestion? If I correctly understand that would solve the main issue that is old results popping up (except for 2.18 that does not have this tag)

pathmapper replied:
I'm also +1 for committing this.

Yep, this should prevent docs build with 

`'outdated': True,`

https://github.com/qgis/QGIS-Documentation/blob/master/conf.py#L146-L148

popping up in search results.

So currently doc builds like `3.28` and `testing` would be shown shown in search results, ideally with the snippet using our HTML `description `meta tag including the QGIS version.

Outdated versions would not be shown in search results, but of course still be available via direct links.

To include the HTML meta tags `robots` and `description` there's a need to rebuild all deployed docs, looks like these are the relevant version (`latest `= 3.`28`).
![grafik](https://user-images.githubusercontent.com/20856381/228144259-aaae390e-c7f6-493d-b7ac-4d5283d7091a.png)

Would this still be possible for all version? If not, I think it still makes sense to merge this PR to include the meta tags for docs where a rebuild is still possible and for future doc builds.

DelazJ replied:
Starting with 3.4, we pulled translated strings in the branch so we have data to rebuild target languages. For 2.18 only english can be rebuilt afaict (unless a pull was done on server side -- but not sure!).
Speaking for the process, it has evolved over the years and I can't tell right now what efforts it may need to put things inline (at least update docker-world, docker-run, makefile files and probably a few things on the server side but that is not really my scope).

kgjenkins commented:
It's great to see this being addressed, and the proposed solution looks promising!

I'm wondering, though, if we might want to shorten the description a bit, since Google and others tend to limit the length of the snippet.  Maybe instead of:
```
Documentation for QGIS 3.18 - 20. Working with Point Clouds
```
we use something like
```
QGIS 3.18 docs - 20. Working with Point Clouds
```

pathmapper commented:
@kgjenkins thanks for your feedback. I don't think we have to shorten the description (there are much longer snippets around, also _docs_ is used as abbreviation for _doctors_) but I do like your suggestion to move the QGIS with version to the beginning (so we could also omit the word `for`):

`QGIS 3.18 documentation: 20. Working with Point Clouds`

That said, I have no strong opinion here.

pathmapper commented:
@DelazJ shoud be good to merge (unrelated test failure).

DelazJ commented:
OK let's then try
